standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveAlterHandler.java

```diff
@@ -187,6 +187,16 @@ public void alterTable(RawStore msdb, Warehouse wh, String catName, String dbnam
             TableName.getQualified(catName, dbname, name) + " doesn't exist");
       }
 
+      String expectedKey = environmentContext != null && environmentContext.getProperties() != null ?
+              environmentContext.getProperties().get(hive_metastoreConstants.EXPECTED_PARAMETER_KEY) : null;
+      String expectedValue = environmentContext != null && environmentContext.getProperties() != null ?
+              environmentContext.getProperties().get(hive_metastoreConstants.EXPECTED_PARAMETER_VALUE) : null;
+      if (expectedKey != null && expectedValue != null
+              && !expectedValue.equals(oldt.getParameters().get(expectedKey))) {
+        throw new MetaException("The table has been modified. The parameter value for key '" + expectedKey + "' is '"
```

Comment:
For my understanding, this only works if a user has table lock while calling alter right?

If user does not have lock, Hive has no internal lock to prevent two users from both getting their expected value and proceeding right?  If so, maybe worth to add a javadoc on this method.

Reply:
Thankfully this is not true.
HMS does these changes using the underlying DB’s transaction.
See:
- https://github.com/apache/hive/blob/84e65538e1009e163d8e81f8645aef9344e24499/standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveAlterHandler.java#L180
- https://github.com/apache/hive/blob/84e65538e1009e163d8e81f8645aef9344e24499/standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java#L558

Reply:
It turns out, that you have been right. We use `READ_COMMITTED` isolation level for our commits.
Because of this, we could commit concurrently, and lose data.

I had to set the isolation level for these transactions to `REPEATABLE_READ` explicitly.

@szehon-ho and @prasanthj could you please check it if you are back from holidays?

Thanks,
Peter

Reply:
Users does not have to do anything. We set the isolation level per transaction automatically from the code. Similarly as we do in the TxnHandler methods where the default isolation level is not enough.

Reply:
This is the relevant code part:
```
msdb.openTransaction(Constants.TX_REPEATABLE_READ)
```

Reply:
Sounds good.